File: sdk/flutter/lib/notifyforge.dart
// NotifyForge Flutter SDK — channel-isolated notification infrastructure.
//
// Usage:
// ```dart
// final nf = NotifyForge(apiKey: 'nf_live_...');
// await nf.push.send({
//   'channel': 'push_android',
//   'target': {'externalUserId': 'user-001'},
//   'payload': {'title': 'Hi', 'body': 'World'},
// });
// ```

import 'dart:convert';
import 'dart:async';
import 'package:http/http.dart' as http;

class NotifyForgeException implements Exception {
  final String code;
  final String message;
  final int status;
  NotifyForgeException(this.code, this.message, this.status);
  @override
  String toString() => 'NotifyForgeException: $code: $message (status $status)';
}

class NotifyForge {
  static const version = '1.0.0';
  final String apiKey;
  final String baseUrl;
  final http.Client _http;
  final Duration timeout;

  NotifyForge({
    required this.apiKey,
    this.baseUrl = 'https://api.notifyforge.dev',
    Duration? timeout,
    http.Client? httpClient,
  })  : timeout = timeout ?? const Duration(seconds: 30),
        _http = httpClient ?? http.Client() {
    if (apiKey.isEmpty) throw ArgumentError('apiKey is required');
  }

  ChannelClient get push => ChannelClient(this, 'push');
  ChannelClient get email => ChannelClient(this, 'email');
  ChannelClient get sms => ChannelClient(this, 'sms');
  ChannelClient get webpush => ChannelClient(this, 'webpush');
  ChannelClient get inapp => ChannelClient(this, 'inapp');
  ChannelClient get webhook => ChannelClient(this, 'webhook');
  ChannelClient get desktop => ChannelClient(this, 'desktop');

  Future<Map<String, dynamic>> request(
    String method,
    String path, {
    Map<String, dynamic>? body,
  }) async {
    final url = Uri.parse('${baseUrl.replaceAll(RegExp(r'/+$'), '')}$path');
    final req = http.Request(method, url)
      ..headers['Authorization'] = 'Bearer $apiKey'
      ..headers['Content-Type'] = 'application/json'
      ..headers['User-Agent'] = 'notifyforge-flutter/$version';
    if (body != null) {
      req.body = jsonEncode(body);
    }
    final streamed = await _http.send(req).timeout(timeout);
    final resp = await http.Response.fromStream(streamed);
    if (resp.body.isEmpty) return {};
    final parsed = jsonDecode(resp.body) as Map<String, dynamic>;
    if (resp.statusCode >= 400) {
      final err = (parsed['error'] as Map<String, dynamic>?) ?? {};
      throw NotifyForgeException(
        (err['code'] as String?) ?? 'http_error',
        (err['message'] as String?) ?? 'unknown',
        resp.statusCode,
      );
    }
    return parsed;
  }
}

class ChannelClient {
  final NotifyForge _client;
  final String _channel;
  ChannelClient(this._client, this._channel);

  Future<Map<String, dynamic>> send(Map<String, dynamic> body) =>
      _client.request('POST', '/api/v1/$_channel/send', body: body);
}

File: sdk/csharp/src/NotifyForge/NotifyForge.cs
// NotifyForge C# SDK — channel-isolated notification infrastructure.
using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.Text.Json;

namespace NotifyForge;

public class NotifyForgeException : Exception
{
    public string Code { get; }
    public int Status { get; }
    public NotifyForgeException(string code, string message, int status) : base($"{code}: {message}")
    {
        Code = code;
        Status = status;
    }
}

public class NotifyForge
{
    private const string Version = "1.0.0";
    private readonly string _apiKey;
    private readonly string _baseUrl;
    private readonly HttpClient _http;

    public NotifyForge(string apiKey, string baseUrl = "https://api.notifyforge.dev")
    {
        if (string.IsNullOrEmpty(apiKey)) throw new ArgumentException("apiKey is required");
        _apiKey = apiKey;
        _baseUrl = baseUrl.TrimEnd('/');
        _http = new HttpClient { Timeout = TimeSpan.FromSeconds(30) };
        _http.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", apiKey);
        _http.DefaultRequestHeaders.Add("User-Agent", $"notifyforge-csharp/{Version}");
    }

    public ChannelClient Push => new(this, "push");
    public ChannelClient Email => new(this, "email");
    public ChannelClient Sms => new(this, "sms");
    public ChannelClient WebPush => new(this, "webpush");
    public ChannelClient InApp => new(this, "inapp");
    public ChannelClient Webhook => new(this, "webhook");
    public ChannelClient Desktop => new(this, "desktop");

    public async Task<JsonElement> RequestAsync(string method, string path, object? body = null)
    {
        var req = new HttpRequestMessage(new HttpMethod(method), $"{_baseUrl}{path}");
        if (body != null)
        {
            req.Content = JsonContent.Create(body);
        }
        var resp = await _http.SendAsync(req);
        var text = await resp.Content.ReadAsStringAsync();
        if (string.IsNullOrEmpty(text)) return default;
        var json = JsonSerializer.Deserialize<JsonElement>(text);
        if (!resp.IsSuccessStatusCode)
        {
            var err = json.GetProperty("error");
            throw new NotifyForgeException(
                err.GetProperty("code").GetString() ?? "http_error",
                err.GetProperty("message").GetString() ?? "unknown",
                (int)resp.StatusCode
            );
        }
        return json;
    }

    public class ChannelClient
    {
        private readonly NotifyForge _client;
        private readonly string _channel;
        public ChannelClient(NotifyForge client, string channel)
        {
            _client = client;
            _channel = channel;
        }
        public Task<JsonElement> SendAsync(object body) =>
            _client.RequestAsync("POST", $"/api/v1/{_channel}/send", body);
    }
}
